Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionEditorToolbar.cs

```diff
@@ -0,0 +1,661 @@
+#if UNITY_EDITOR
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEditor;
+using UnityEditor.IMGUI.Controls;
+using UnityEditorInternal;
+using UnityEngine.Experimental.Input.Utilities;
+
+////TODO: better method for creating display names than InputControlPath.TryGetDeviceLayout
+
+namespace UnityEngine.Experimental.Input.Editor
+{
+    /// <summary>
+    /// Toolbar in input action asset editor.
+    /// </summary>
+    /// <remarks>
+    /// Allows editing and selecting from the set of control schemes as well as selecting from the
+    /// set of device requirements within the currently selected control scheme.
+    ///
+    /// Also controls saving and has the global search text field.
+    /// </remarks>
+    /// <seealso cref="InputActionEditorWindow"/>
+    [Serializable]
+    internal class InputActionEditorToolbar
+    {
+        public void OnGUI()
+        {
+            EditorGUILayout.BeginHorizontal(EditorStyles.toolbar);
+            DrawSchemeSelection();
+            DrawDeviceFilterSelection();
+            if (!InputEditorUserSettings.autoSaveInputActionAssets)
+                DrawSaveButton();
+            GUILayout.FlexibleSpace();
+            DrawAutoSaveToggle();
```

Comment:
Yeah, I think that could have been done better.

It evolved out of a decision to have JSON text formats for all data consumed by the system. Which ended up putting the editor on a track where it saved out in that format.

I think there's also some other complications that I have mostly forgotten about. The data you mention above is all Unity native data. But we're sitting in a package and IIRC there were some problems there. But regardless of what it was, I'm thinking it probably would have been solvable.

FWIW I think the new shader graph stuff works in a similar way.

>can we add a (very) visual dirty indicator to the window?

What do you suggest? The current one IMO is pretty visible but could well be it should be more prominent.

Reply:
>The current one IMO is pretty visible 

Heh, I did not notice that was there, but I guess that's because I use the development editor build, which draws the UI update widget over the corner of the window, partially covering it. So maybe that's fine (I guess ideally, I'd want a dot in the red window close button as per macOS conventions, but i guess we cannot do that without editor-level support).

But maybe we should reconsider just always saving (and removing the explicit save UI) to match unity conventions. The fact that all the examples I mentioned "is all Unity native data." should not make any difference to the user here.